### ISSUE TITLE
Data file with several power modules - power module name and non unique channel names

### ISSUE THREAD

**Issue by yellow-submarine-on-tour** [costerwi/dwdatareader#34]
Hi,

we are using the power module function a lot, with several AC and DC power modules in one measurement. The problem is, that the dwdatareaderlib-API gives only the channel name (like U_L1, U_L2, P_H1, ...), but not the related power module name itself. Hence there are several channels with the same name

This has several implications:
- `__getitem__` gives only the first occurrence of a channel with the given name. We cannot, for example, get the second occurrence of the name
- `__iter__` also returns the name, which results in the same problem. One cannot iterate over DWFile and then index with the channel name you get in the iteration
- Indexing by number results in two nested for loops (we have data files with >400 channels and > a few GB, this is just super long)
- The only way you can find out more about the channel is by requesting the channel-XML. But this only gives you the information that the channel is part of a power module with its according index, not the name of the module itself

Our solution/dirty hack is:
- Extract header file from data file
- Analyze this XML file, go through all math channels and then through all power modules
- Make a list of all power modules with name and index
- Iterate through all channels by index, check if name (like U_L1) and power module index matches with our name/index-list
- If yes, read data
- When done, delete XML header file

The Dewesoft support knows the problem and affirmed that our solution is at the moment the only way to get the power module name. And it sounded that there won't be an update in the dwdatareaderlib in the near future that solves the problem in an elegant way. So I wanna start a discussion if and how we can integrate it in the dwdatareader

One point is, how to get channel data when the name is not unique. Then, handling huge files with a lot of channels. Finally and most important, how we can implement something like `DWChannel.get_math_module_number()`, `DWChannel.get_math_module_type() ` and `DWChannel.get_math_module_name() `?

Best wishes
/chr

**Comment by costerwi** [comment]
The indexing has always been ugly. I assume ch.index is unique? Would it improve things if `__iter__` returned that index instead of the channel name? Potentially we could make performance improvements to `__get_item__` when searching for a channel by index.

Unfortunately, I don't know what a power module is and I don't find "power" or "module" when I search the latest DLL documentation so you'll have to help with that.

**Comment by yellow-submarine-on-tour** [comment]
The power module is a plugin one have to buy additional which can calculate voltage and current RMS values, active and reactive power, harmonics, ... from current and voltage measurements (c.f. https://manual.dewesoft.com/x/setupmodule/modules/electrical/power)

The resulting channels are async variables (sync to the grid frequency, but as the frequency is not constant, the channel rate is not constant). Beside that, channels from the power module are the same as math channels one creates manually. They are "reduced" from the sample rate (typically a few kHz, power channel rate is per default 10 periods, so 5Hz/6Hz), but they are not reduced in a way like the storing option "slow" with reduced rate (matching the `def reduced()` in the lib).

FYI, the channel_xml of a power module channel is:
```xml
<OutputChannel Descr="FrequencyErrorStatus" Index="0">
...
	<Settings>Math 6 (Power analysis)</Settings>
</OutputChannel>
```
And the XML one gets with `export_header()` is:
```xml
<DataFileSetup>
	<System Name="Local">
		<DewesoftSetup>
			<Math>
				<Math Type="Formula" Index="5">
				...
				</Math>
				<Math Type="Power" Index="6">
					<Name>Batt_Inv_AC</Name>
				</Math>
			</Math>
		</DewesoftSetup>
	</System>
</DataFileSetup>
```

For me, it would help a lot, if `__iter__` returns the index, but I don't know how many other people use the lib and if that change breaks their code. Or I would even prefer not to yield a `name` or `index`, but the channel `ch` itself.

What I do now as a workaround, is
```python
    for channel in f.channels:
        channel = f[channel.index]
```
which seams absurd to reasign again. But without this `channel = f[channel.index]` the call of `channel.channel_xml` gives `dwdatareader.DWError: error in DLL` exception. Also just doing a `for i in range(n)` and then `channel = f[i]` is not working. I assume it has something to do with ctypes and this line `self.channels = (DWChannel * nchannels)()`, but I'm not an ctype expert.

**Comment by costerwi** [comment]
The channel you get from iterating through `f.channels` should be no different from `f[channel.index]` unless the `channel.index` is not unique. The `f[channel.index]` will return the first channel it finds with that index. I am able to run the following against the sample d7d file:
```python
for channel in f.channels:
   channel.channel_xml
```
The fact that you also cannot index by `range(n)` is another clue that there is not a 1:1 relationship between `channel.index` and its position within the DWChannel array. This might be an interesting test to see how far off they are:
```python
for n, channel in enumerate(f.channels):
   if n != channel.index:
      print(n, channel.index, channel.name)
```
It seems plausible that your power module channels are special and don't store any data in the .d7d file since it can be calculated later from the other data. Maybe calling the normal Dewesoft DLL routines with them will cause DLL errors.

**Comment by yellow-submarine-on-tour** [comment]
> The channel you get from iterating through `f.channels` should be no different from f[channel.index]

The thing is, that iterating through `f.channels` gives me the same numerical data than `f[channel.index]`, but as I wrote, iterate over 'f.channels' and then calling `channel.channel_xml` gives an exception `"dwdatareader.DWError: error in DLL"`. This makes no sense for me, I have no idea why. Maybe its a bug in the DLL. Maybe I have to read data first, but this is not feasible, as loading takes a lot of time, just to then decide to reject it.

> This might be an interesting test to see how far off they are:
```python
        for n, channel in enumerate(f.channels):
            if n != channel.index:
                print(n, channel.index, channel.index-n, channel.name
```
gives
```
106 107 1 Status
107 108 1 P_L1
108 109 1 Q_L1
109 110 1 S_L1
...
120 121 1 E_P_L1
121 123 2 Status
122 124 2 PF_L1
123 125 2 I_rms_L1
...
156 158 2 S_H1
157 159 2 QH_H1
158 160 2 E_P
159 161 2 E_P+
160 162 2 E_P-
161 164 3 U_ave
162 165 3 I_ave
163 166 3 P_ave
164 167 3 E_P
...
241 244 3 E_S
242 245 3 E_P+
243 246 3 E_P-
244 248 4 P_batt_inv_tot
245 249 4 AcqDisplayTime
246 250 4 AcqAcquireDataTime
247 251 4 AcqStoreTime
248 252 4 SyncTimeTracking
249 253 4 Temperature
250 254 4 FanSpeed
251 255 4 Temperature
252 256 4 FanSpeed
253 257 4 Temperature
254 258 4 FanSpeed
```
which BTW shows, that also meta informations like HW module temperature names are not unique

>  and don't store any data in the .d7d file since it can be calculated later from the other data

We only store the power module channels, as otherwise the files would be just toooo big.

**Comment by costerwi** [comment]
My theory that you're getting a different channel with the same index doesn't explain the error. In the end the Python is simply calling `DLL.DWGetChannelProps` with the exact same `channel.index` in any case. I can't explain why the DLL call would work sometimes but not others on the same index but we did recently update the DLL which may make a difference #33.

Have you identified which channel is having a problem with `channel.channel_xml`? I assume most of them do not have the error. Maybe there is a clue there.

The debug data you shared shows that some `channel.index` are skipped. For example, 106 and 122 are missing. I think this is why `range(n)` was not working for you.

**Comment by yellow-submarine-on-tour** [comment]
I prepared a longer reply/update on this, but I just realized, that I never postet it. Sorry about this...!

To sum it up:
- I was in contact with Dewesoft support regarding this
- They promised a solution for this. After some delays, they published a new version of data reader lib (https://download.dewesoft.com/list/developers/dewesoft-data-reader-library, v4.2.0.21 from 2021-11-29)
- A new property is introduced, the channel long name. That includes Power module name, statistic package, ...

I added some code to the lib:
```python
class DWChannelProps():
    DW_DATA_TYPE = 0
    ...
    DW_CH_XMLPROPS_LEN = 10
    DW_CH_CUSTOMPROPS = 11
    DW_CH_CUSTOMPROPS_COUNT = 12
    DW_CH_LONGNAME = 13
    DW_CH_LONGNAME_LEN = 14
```

```python
    @property
    def channel_long_name(self):
        return self._chan_prop_str(DWChannelProps.DW_CH_LONGNAME, DWChannelProps.DW_CH_LONGNAME_LEN)
```

This gives for a test file with two DC power modules:
```
PVS_S1/U_ave
PVS_S1/I_ave
PVS_S1/P_ave
PVS_S2/U_ave
PVS_S2/I_ave
PVS_S2/P_ave
```

**Comment by costerwi** [comment]
Closed by #47
